src/pages/blog/optimizing-emotion.mdx

```diff
@@ -0,0 +1,293 @@
+  - Emotion
+  - Performance
+---
+import { Card } from "../../components/Card";
+
+## What is Emotion?
+
+Emotion is a series of NPM packages that allow for extensible CSS-in-JS dynamic style construction, all under the `@emotion/*` organization. It is used as the styling engine for several popular component libraries including `@mui/*` and `@chakra-ui/*`.
+
+With the `@emotion/css` library, you can generate styles under a randomized className at runtime which can allow it to be used effectively with or without any framework. The `@emotion/react` library has special integrations with React that allow for additional optimizations and a `React.Context` driven theme system. Finally, the `@emotion/styled` library allows for the creation of React components by only specifying an HTML element name and it's associated styles (a pattern borrowed from the `styled-components` library). There are actually several other libraries in the Emotion ecosystem, but the rest are mostly underlying core libraries, compiler plugins, or testing tools.
+
+If used correctly, Emotion can be a great tool for constructing highly dynamic applications and is perfomant enough for the task. It also has the DX benefit for React applications that is already found in most alternate framework templates in colocated styles. For example, Vue, Svelte, Astro, and others all allow for component scoped styles in the same file as the component logic and template. This makes for a great developer experience and Emotion brings this to React.
+
+## So what is the problem?
+
+Emotion can be slow. It is not always slow and for what it has to do, Emotion is actually fairly performant. However, Emotion inherently has to do more runtime work than using raw CSS, CSS modules, a CSS preprocessor (e.g. [Sass/SCSS](https://sass-lang.com/), [Less](https://lesscss.org/), or [Stylus](https://stylus-lang.com/)), or another zero-runtime solution for styles (e.g. [TailwindCSS](https://tailwindcss.com/), [Compiled](https://compiledcssinjs.com/), or [Vanilla Extract](https://vanilla-extract.style/)). Raw CSS requires no extra processing. CSS modules, CSS preprocessors, and zero-runtime solutions all require some extra processing at build & dev time, but the resulting CSS is static and can be minified & cached so the client gets optimal performance with these solutions. Emotion, on the other hand, has to do all of the processing at runtime and does comparably similar work to both CSS modules _and_ a CSS preprocessor.
+
+## Just how much work is Emotion doing?
+
+Just like at the build time of a CSS preprocessor, Emotion has to parse the style passed into it. Emotion _does not_ simply pass its input strings into a `<style>` tag, it must parse the style strings into an object tree. The reason your style strings must be parsed by Emotion before the browser is because Emotion offers SCSS-like nesting syntax as it uses [`stylis`](https://www.npmjs.com/package/stylis) as it's underlying stylsheet language.
```

Comment:
TIL about Stylis

Reply:
I learned about it this week. I did a lot of deep-diving into the source contributor docs of Emotion to write this article 😅